doc/hare-dtm-hld.md

```diff
@@ -0,0 +1,614 @@
+bootstraps, mainly by the RC leader when it is elected. An epoch is a counter
+that is updated only by RC leader and used to record the cluster wide HA
+events with respect to Motr processes.
+When a Motr process transitions to ONLINE the corresponding hax process
+receives an event (M0_CONF_HA_PROCESS_STARTED) from the Motr process.
+Hax then creates or updates the `<proc-fid>/current_epoch:epoch#` KV.
+Additionally, Hax creates `failvec/<proc-fid>/<Epoch>:{}` KV where the
+relative failure events are recorded. Epoch number initially is set to 0 and
+the corresponding failure vector KV is also initialized to an empty list.
+failure events relative to this Motr process since the process started are
+then atomically recorded at the failure vector KV for a given epoch.
+Thus all the device failures and the process failure itself are recorded with
+respect to this epoch. When a Motr process restarts the current epoch number
+(`<proc-fid>/current_epoch:epoch#`) is updated and a new epoch KV is created
+by the corresponding hax process.
+Following sequence diagram illustrates the same,
+<!--
+@startuml
+autonumber
+systemd -> haxN: Start
+systemd -> MotrN: Start
+systemd -> consul: Start
+MotrN -> haxN: Entrypoint request
+haxN -> MotrN: Entrypoint reply
+haxN -> MotrN: DTM recovery start if required
+MotrN -> haxN: Recovery complete
+MotrN -> haxN: notify M0_CONF_HA_PROCESS_STARTED
+haxN -> consul: fetch and update or create "proc-fid/epoch"
+haxN -> consul: create "failvec/proc-fid/epoch:{}"
+@enduml
+-->
+![img](ha-epoch.png)
+
+### Failure order
+Hare records all the failure events with respect to Motr processes in the
+cluster in the order as they are received. It is important that the failures
+are arranged such that it is efficient to compile a list and prune the same as
+required. Failures can be recorded in order as follows,
+
+Consul key prefix | value
+--- | ---
+failvec/proc-fid/epoch/nr_failures | nr_failures
+failvec/proc-fid/epoch | Dictionary of configuration objects and ha state
+
+e.g.
+failvec/proc-fid/epoch:{<conf-obj-fid>:FAILED}
+
+When a hax process receives a failure event for a Motr process, it reads
+the last ONLINE event epoch and adds a new entry to event queue in-order
+to update failure vector.
+
+#### Failvec update
+Hare event and broadcast queue mechanism is leveraged here. When a hax process
+receives a failure event, it creates an event and enqueues it into the event
+queue (EQ). RC leader then processes the event. It updates the failure count
+with respect to that epoch and creates a new failure vector entry as mentioned
+above. While updating the failure vector, it is possible that there are duplicate
+event queue entries for the same event (as the failure event is received on
+multiple nodes). In order to handle this duplication, RC leader checks if the
+corresponding failure event is already added to the failure vector for the
+given epoch and discards the duplicate events if already recorded.
```

Comment:
>How RC will know that two events are duplicate ones? Will it compare timestamps (assuming that if two events happened quite close to each other then they represent just the same fact)?

>I guess that different subsystems will produce events of different types (e.g. "I didn't manage to read block device", "disk is full", "disk is unmounted") while they still represent the same failure and the same events. Will we need to interpret all of them at Hare level?

The events will be purely from hare perspective, meaning, if a device or a process fails, an event is posted to RC. Now, Now device failure events are received only by RC but process failure events will be received by all the hax processes. So given the tuple <fid, state>, RC creates a key if not already exists. While processing subsequent events, RC would check if the key for the given tuple exists in the failvec, if yes, then ignores. Note that, only the failure events are received by all the haxs but subsequent transitions are done only by one hax process that manages the corresponding Motr process. Will add this to design highlights.

>The problem is probably worse. All the duplicate events will be put to EQ. So "checking" for duplicates will happen while reading out the event queue. In this case RC will either need to look ahead into EQ (which doesn't look like quite normal queue then) or RC will need to store the history of "already processed events".

I think RC just needs to look for the <fid, state> tuple in the failvec in case of failures. If already applied then it need not reapply. 

I also thought about using locking but that too did not seem very beneficial as we already have RC and queue mechanism which also helps us to handle hax restarts and process the events in a synchronous manner which is important to apply the events in some order. One thing we can do is to instead of RC processing duplicate events, we can try and not post duplicate events itself but that will still be tricky and I see races in that too.